### ISSUE TITLE
F-Droid reproducible builds

### ISSUE THREAD

**Issue by JohnBetaro** [OxygenCobalt/Auxio#370]
The title is clear, if you can publish your app on play store or izzyondroid f-droid repository

i don't trust the apps that are built by f-droid for security reasons, i prefer to get apps that are signed with the developer's own keys, please consider my thoughts you will make make things even more accessible and secure.

**Comment by OxygenCobalt** [comment]
- I think IzzyOnDroid does not accept apps that are already on F-Droid to avoid build conflicts.
- I could work on making the F-Droid build reproducible, but that takes a good amount of effort on my end as my generated APKs and F-Droids APKs must match byte-to-byte.
- I am also planning to publish on Accrescent [#284] eventually, which has native support for using my own signing keys. However, Accrescent is early in development and I'm awaiting for it to look at least somewhat complete before publishing.

Number 2 seems like the most reasonable option.

**Comment by JohnBetaro** [comment]
I see LibreTube provided by both izzyondroid and also f-droid, but as you said do what you see the best.

**Comment by OxygenCobalt** [comment]
Hm. Guess I was just assuming what happened when I got onto F-Droid initially (Auxio was de-listed from IzzyOnDroid).

**Comment by JohnBetaro** [comment]
I opened an issue on izzyondroid GitLab page and asked the owner to add the app and I told him it's already on f-droid but we want to get it straight from the main developer, we'll see his respond.

**Comment by JohnBetaro** [comment]
Hi, I'm here again to ask you for your approval to add the app to izzyondroid repository because the owner asked me if the developer of the app accepts adding his app even there will be a conflict (having 2 different signatures), please respond in the nearest time possible.

**Comment by OxygenCobalt** [comment]
I guess I can't right now since it would be a breaking change. Lot's of people that use IzzyOnDroid and F-Droid simultaneously would suddenly have a new APK with conflicting changes. 

In fact, now that I think about it, I don't even know if I can make the build reproducible on F-Droid itself without that being a breaking change. I would have to ask the developers on if there is any migration pathway.

**Comment by JohnBetaro** [comment]
Actually I don't get where the problem is exactly, me personally I use droid-ify and i can choose which variant of the app I want to install and there will no problem at all, so please if you can explain and make it easy on me to understand.

**Comment by OxygenCobalt** [comment]
Not everyone uses droid-ify, I have no idea how the F-Droid client handles this.

**Comment by JohnBetaro** [comment]
![Screenshot_20230224-051706.png](https://user-images.githubusercontent.com/114379310/221090643-6e5d6439-656e-4c7c-b615-fe7bc626057b.png)

https://user-images.githubusercontent.com/114379310/221090715-271d7765-380d-491d-b0fc-fe781234ff4c.mp4

I installed the official f-droid client and as you see :

1- The first screenshots confirms that you can you can choose which repository you want to get your app from

2- The second is a video that confirms f-droid will choose the right apk (with the same signature as installed) when you want to do an update (I installed my app from izzyondroid then I disabled the repository)

**Comment by IzzySoft** [comment]
> i don't trust the apps that are built by f-droid for security reasons

Which I find quite strange. The process on F-Droid is quite transparent; I don't see any security reason to not trust the signature F-Droid provides. My guess is you've read one of those misleading articles claiming things like "all apps are signed with the same key" and "they keep their key available online"? Be assured, that's pure :hankey: Each app gets its own signing key (on request and where needed, a shared key can be created for two apps of the same author which need it to work properly), and signing happens with a secure keystore on an air-gapped server.

As the title of your issue is "reproducible builds": Here I can agree that would add one level. In that case, the app would be built by F-Droid the very same way as it is now. The difference is when it comes to signing: instead of signing the APK with its key generated by F-Droid, it will be compared against the APK built by the author, and if the two are identical (apart from the authors signature of course), the APK signed by the author goes to the index. So you have the added transparency and checks by F-Droid (making sure the APK really corresponds to the code and as checked against "unwanted ingredients") with the confirmation of both, F-Droid **and** the author. While otherwise it would be F-Droid **or** the author, depending on which signature.

> I think IzzyOnDroid does not accept apps that are already on F-Droid to avoid build conflicts.

Not build conflicts, but rather confusion due to signature conflicts. Auxio started in my repo on 2021-02-18, then half a year later (2021-07-05) was established on F-Droid, and given the usual decent overlap was removed from my repo 10 days later (which is the default action if not explicitly requested otherwise **by the app's author**). Exceptions are possible, but should be exceptions, so there should be a good reason for it. For one, there's the signature conflict – and for the other, my repo runs on personal space (and basically is a "one-man-show") so resources are limited. With reproducible builds and an app having a fast pace (e.g. daily releases while F-Droid would just spit out new releases every 2..3 days, but the releases being e.g. security relevant) it's a given that I agree (as long as the APK size stays within limits). In other cases, you'll need to convince me why it is important – especially when adding it while it's already available at F-Droid.

> The first screenshots confirms that you can you can choose which repository you want to get your app from

And how many people do you think are aware that they should scroll down there for each app and check from where it's installed? My guess is the majority uses the official client and is **NOT** aware of that. There's a slight difference between @OxygenCobalt offering their own repo which people have to add explicitly (which they of course only do if they want their app this way) – or people adding my repo for some extra apps available only there and then "accidentally" get Auxio from me while assuming they get it from F-Droid.

So TL;DR: can be done if there are good reasons – but there should be good reasons.

(PS: I've chimed in because @JohnBetaro [requested inclusion](https://gitlab.com/IzzyOnDroid/repo/-/issues/351) of your app with my repo)

**Comment by JohnBetaro** [comment]
I'll leave [https://privsec.dev/posts/android/f-droid-security-issues/](this article) here i think all people should read it, as much as I like f-droid and its purpose but their bad security practices made me question anything I get from their repository.

**Comment by JohnBetaro** [comment]
I'll leave [this article](https://privsec.dev/posts/android/f-droid-security-issues/) here i think all people should read it, as much as I like f-droid and its purpose but their bad security practices made me question anything I get from their repository.

**Comment by IzzySoft** [comment]
Thought so. You need to understand the concepts to understand the critique. While the author indeed tries to take a neutral/objective stance, it's still required that the **reader** understands the underlying principles or the meaning is lost. For example, do you understand the backgrounds for targetSDK, why that's needed for Playstore but not necessarily for F-Droid? The article does not tell (it even forgets to mention implications), so you have to take it by the word (or read on [some background here](https://social.librem.one/@eighthave/109545306371981727)). While well intended, if you don't have the background to understand the implications, it's easy to come to the wrong conclusions. As you possibly have, at least in parts :wink:

PS: No, I won't discuss all details of that article – I lack the time to do that. But it seems the author didn't contact F-Droid for clarification on several items. Examples: the whole build infra was in an overhaul over the past 2 years, and the new one (running on Debian Bullseye) was put to production around the time when the article was published. At the same time, the client still is in a complete overhaul which is not yet finished (parts you can already see in the alpha releases, other parts are in a separate branch as they are not yet completed). While Playstore tries to hide implications (tracking libraries, excessive permissions etc), F-Droid uses a transparency approach (listing all permissions, not permitting proprietary libraries and showing if some "unwanted features" are present) which I feel misrepresented in that article – and that the Android permission model is far from being perfect you can find confirmed by security researchers as well (see e.g. [here](https://www.kuketz-blog.de/das-android-berechtigungsmodell-ein-perfides-konstrukt/) in German) – especially considering runtime permissions as "the solution" (ever tried denying an app the INTERNET permission?).

But `EOF` and `</OT>`.

**Comment by JohnBetaro** [comment]
Alright then I have to thank you for guiding me, I'll be sure to do my own research on this topic and not jump to conclusions very quickly.

Thank you again ❤️

**Comment by IzzySoft** [comment]
Thanks John! Should you still think Auxio should be re-added to my repo for some reason, it's up to @OxygenCobalt to say if he wants it that way. After all, it won't be you or me people will complain to when confused :wink:

**Comment by OxygenCobalt** [comment]
@IzzySoft I want to see if you know: What would happen if I were to make my builds reproducible on F-Droid? Would F-Droid be able to provide dual builds with the my signature and F-Droid signatures to maintain backwards compatibility?

**Comment by IzzySoft** [comment]
>  What would happen if I were to make my builds reproducible on F-Droid?

If that could be achieved, the confusion part would be minimized :wink:

> Would F-Droid be able to provide dual builds with the my signature and F-Droid signatures to maintain backwards compatibility?

Not exactly my area of experience. Would this be a new app to be introduced to F-Droid it would be clear: only the APK signed by you would be shipped, and that's it. This definitely works, there are close to 100 such apps listed. But if you want to enable reproducible builds later on, while your app is already listed, things become a bit more complicated. By theory, there are the following possible approaches I could think of:

* you add a separate build (flavor) with a different `applicationId`. That would be an entirely new app and thus handled as such – but existing users had to switch by uninstall/reinstall with some migration.
* you add a separate build flavor using a different `versionCode` (basically similar to per-ABI builds) and define reproducibility per-build-block. So e.g. there'd be a version 2.0.0 signed by F-Droid using `versionCode: 200` (the one currently used), and another (new one) 2.0.0 built reproducibly using `versionCode: 201` and carrying your signature. That way, existing users (F-Droid signed APK) would get offered the 200 (highest one with a matching signature), while new users would get offered 201 (highest one available). This **could** work, but I'm not sure if it's "feasible" at F-Droid (I didn't see it used that way). Worth asking, though.
* dual signing (i.e. the APK comes with both signatures). This should be possible with v3 signing AFAIK ("key rotation") and is **planned** at F-Droid (to my knowledge), but not yet available (as the framework does not yet support it). No ETA for this either.

> Or would it be a (likely) breaking change?

That would happen if you'd change to just a single reproducible build, signed with only your key (as then existing users of the F-Droid signed APK could not upgrade without uninstall/reinstall – basically like the first bullet-point above).

> If it's not breaking, then I'll do it. Only remaining concern would be making sure that my key isn't lost.

Your key entirely stays with you. To put the "simple" process in short terms:

* you provide a tag with your signed APK (built from the commit the tag points to) attached to it
* F-Droid builds the APK from that very same commit itself, then fetches your APK to compare
* the "signing block" from your APK is copied to the APK built by F-Droid
* `apksigner verify` is run on that "patched" APK. This can only succeed if the APK itself was binary-identical. So if it succeeds, the APK built by F-Droid is thrown away and the one you provided will be shipped. If it fails… well, the build fails and we'll have to fix it.

For some more detailed reading:

* [HOWTO: diff & fix APKs for Reproducible Builds](https://gitlab.com/fdroid/wiki/-/wikis/HOWTO:-diff-&-fix-APKs-for-Reproducible-Builds)
* F-Droid Blog article: [Towards a reproducible F-Droid](https://f-droid.org/2023/01/15/towards-a-reproducible-fdroid.html)
* [reproducible-builds.org article](https://reproducible-builds.org/de/news/2022/06/24/supporter-spotlight-hans-christoph-steiner-f-droid-project/)
* and finally the "official documentation" in [F-Droid Docs](https://f-droid.org/docs/Reproducible_Builds/)

**Comment by OxygenCobalt** [comment]
Okay. Seems like this isn't exactly feasible for me at the moment then. I want to reduce the amount of complexity in my build recipes anyway so I don't have to constantly make fdroiddata PRs for every new release. I think I'd wait until multi-sig is supported (whenever that is) and then revisit. I still point to #284 for people like OP who prefer cathedral-style security models, as I think Accrecent will serve that demographic a lot better when it reaches 1.0.

**Comment by JohnBetaro** [comment]
@IzzySoft For now i'm using the apk from F-Droid until I see a significant risk that can make me go back to the GitHub version, and the app does not have access to my sensitive files, only the audio files, I think I can wait but it will be better to have it directly from @OxygenCobalt if he managed to solve the issue.

**Comment by OxygenCobalt** [comment]
Seems like that only occurs if you add a signature to your metadata @IzzySoft. This requires me to manually upload my signature to `metadata/signatures`, which then requires me to disable auto-updating (Again, don't want to do that since filing a PR every release is super tedious). Really confused why it does not occur with binary comparison.

**Comment by IzzySoft** [comment]
Thanks @JohnBetaro! And to not "hijack" Alexander's attention: if you have further questions on this topic, let's discuss them in the issue over at my repo. Also feel welcome to discuss it in the [F-Droid Forum](https://forum.f-droid.org/) (no worries about neutrality: it's not just "F-Droid staff" there, but anyone can discuss there openly; though you might get some interesting insights from other F-Droid team members as well: transparency is very important for us (yes, I'm a member of the F-Droid team, too), so this can help you form your own opinion on that getting the point of "both sides").

@OxygenCobalt I just run a "simple binary repo" and take the developers' APK as-is. I'm not even an Android dev, so my understanding of programming, compiling etc. is rather basic. The process of reproducible builds was much enhanced in the past few months (in fact most of the apps using this were added only very recently), so if I understood correctly signatures can be extracted from your APK directly as well. As pointed out, those questions are better directed to those with real in-depth understanding. Mine is much to basic for that :wink:

And a huge thanks to you for caring and taking your time for such a lengthy discussion, much appreciated!! :star_struck:

**Comment by OxygenCobalt** [comment]
[Started a new thread here](https://forum.f-droid.org/t/reproducible-builds-any-plans-to-publish-dual-apks-with-binaries-build-recipe-attribute/21437)

**Comment by IzzySoft** [comment]
Thanks!
